### PR TITLE
16160 - EMAILER - Special Resolution and Correction

### PR DESCRIPTION
**Pull request by Jxio**
*Issue #:* [/bcgov/entity/16160](https://github.com/bcgov/entity/issues/16160)

*Description of changes:*
2. Send email notification when Special Resolution and Correction have been filed
3. Filer fix - coopAssociatiionType and signing_date for special resolution correction

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: queue_services/entity-emailer/src/entity_emailer/email_processors/correction_notification.py [hunk 1]
```diff
@@ -45,6 +47,10 @@ def _get_pdfs(
         'Authorization': f'Bearer {token}'
     }
     legal_type = business.get('legalType', None)
+    is_cp_special_resolution = (
+        legal_type == Business.LegalTypes.COOP.value and
+        filing.filing_json['filing']['correction'].get('correctedFilingType') == 'specialResolution'
+    )
```
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/email_processors/correction_notification.py:53]: so do you need to handle the correction scenario where you do a correction on the original SR correction like you did for the filer?

Comment (Jxio) [reply queue_services/entity-emailer/src/entity_emailer/email_processors/correction_notification.py]: yes, thanks for point out, I just added a is_special_resolution_correction helper function to trace original sr filing.

File: queue_services/entity-emailer/src/entity_emailer/email_processors/special_resolution_notification.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    # get attachments
+    if status == Filing.Status.PAID.value:
+        pdfs = get_paid_pdfs(token, business, filing, leg_tmz_filing_date, leg_tmz_effective_date)
+    if status == Filing.Status.COMPLETED.value:
+        pdfs = get_completed_pdfs(token, business, filing, name_changed)
+
+    # get recipients
+    identifier = filing.filing_json['filing']['business']['identifier']
+    recipients = []
+    recipients.append(get_recipient_from_auth(identifier, token))
```
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/email_processors/special_resolution_notification.py:70]: I think you might have an issue here.  From the SR email requirements, you need to email the registered office and the completing party.

The registered office email is what you are retrieving via `get_recipient_from_auth(identifier, token)` so I think we are good there.

But the completing party is never saved to the filing json even though the SR designs have it auto-populated.  i.e. we don't have the completing party to send to currently
image

I think the business schema for SR needs to be updated to include a `parties` block so the completing party can be added when the filing json is saved.  I'm assuming some FE work will also need to be done to ensure the completing party is added to the filing json.

File: queue_services/entity-emailer/src/entity_emailer/email_templates/CP-SR-CRCTN-COMPLETED.html [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            [[business-info.html]]
+            <p>The following documents are attached to this email:</p>
+            <ul class="outputs">
+                <li>Special Resolution</li>
+                {% if name_changed %}
+                    <li>Certificate of Name Change</li>
+                {% endif %}
+                {% if filing['rulesInResolution'] %}
+                    <li>Certified Rules</li>
+                {% endif %}
```
Comment (argush3) [inline queue_services/entity-emailer/src/entity_emailer/email_templates/CP-SR-CRCTN-COMPLETED.html:35]: I think `if filing['rulesInResolution'] %}` should be `if filing['rulesFileKey'] %}`.  "Certified Rules" should only show when a new rules pdf is uploaded right?

Comment (Jxio) [reply queue_services/entity-emailer/src/entity_emailer/email_templates/CP-SR-CRCTN-COMPLETED.html]: you're right, I just updated

File: queue_services/entity-emailer/src/entity_emailer/email_processors/special_resolution_notification.py (hunk 2 above)
Comment (vysakh-menon-aot) [reply queue_services/entity-emailer/src/entity_emailer/email_processors/special_resolution_notification.py:70]: This should solve the issue https://github.com/bcgov/lear/blob/main/queue_services/entity-emailer/src/entity_emailer/email_processors/filing_notification.py#L276

Comment (Jxio) [reply queue_services/entity-emailer/src/entity_emailer/email_processors/special_resolution_notification.py:70]: I just look at SR base filing, completing email send with ['filing']['header']['documentOptionalEmail'], I think recipients added as:
```
    if filing.submitter_roles and UserRoles.staff in filing.submitter_roles:
        recipients.append(filing.filing_json['filing']['header'].get('documentOptionalEmail'))
 ```
 I'll modify business-schema to include parties for both SR and SR correction

Comment (vysakh-menon-aot) [reply queue_services/entity-emailer/src/entity_emailer/email_processors/special_resolution_notification.py:70]: Refer the above link. You don't need to modify schema

Comment (Jxio) [reply queue_services/entity-emailer/src/entity_emailer/email_processors/special_resolution_notification.py:75]: @vysakh-menon-aot Thanks! I updated.

**Review by vysakh-menon-aot — APPROVED** [review]

File: queue_services/entity-emailer/tests/unit/email_processors/test_cp_special_resolution_notification.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                email = special_resolution_notification.process(
+                    {'filingId': filing.id, 'type': 'specialResolution', 'option': status}, TOKEN)
+                if status == 'PAID':
+                    assert email['content']['subject'] == LEGAL_NAME + \
+                        ' - Confirmation of Special Resolution from the Business Registry'
+                else:
+                    assert email['content']['subject'] == \
+                        LEGAL_NAME + ' - Special Resolution Documents from the Business Registry'
+
+                assert RECIPIENT_EMAIL in email['recipients']
```
Comment (argush3) [inline queue_services/entity-emailer/tests/unit/email_processors/test_cp_special_resolution_notification.py:58]: can we add a check for `USER_EMAIL_FROM_AUTH` as well?

Comment (Jxio) [reply queue_services/entity-emailer/tests/unit/email_processors/test_cp_special_resolution_notification.py:58]: yes, just added

**Review by argush3 — APPROVED** [review]

**Pull request merged**